### PR TITLE
Fix intersect error reported in #41096

### LINKED ISSUE #41096: type-intersection env may fail merging of multiple value matches
When I was developing my own package, I came across a weird bug of Julia on the control flow using the ``if``-``else`` statement. The minimal reproducible example is as follows:
```julia
struct Modulate{M<:Union{Function, Val{true}, Val{false}}, id}
    modulate::M
    Modulate(id::Symbol, modulate::Function) = new{typeof(modulate), id}(modulate)
    Modulate(id::Symbol, modulate::Bool=true) = new{Val{modulate}, id}(modulate|>Val)
end
@inline ismodulatable(modulate::Modulate) = ismodulatable(typeof(modulate))
@inline ismodulatable(::Type{<:Modulate{Val{B}}}) where B = B
@inline ismodulatable(::Type{<:Modulate{<:Function}}) = true

mutable struct Term{I, M<:Modulate}
    modulate::M
    Term{I}(modulate::Modulate) where I = new{I, typeof(modulate)}(modulate)
end
@inline ismodulatable(term::Term) = ismodulatable(typeof(term))
@inline ismodulatable(::Type{<:Term{I, M} where I}) where M = ismodulatable(M)

function newexpand(gen, name::Symbol)
    flag = ismodulatable(getfield(gen, name))
    println(flag)
    if flag
        println("flow for true.")
    else
        println("flow for false.")
    end
end

t = Term{:t}(Modulate(:t, false))
μ = Term{:μ}(Modulate(:μ, false))
U = Term{:U}(Modulate(:U, false))

newexpand((t=t, μ=μ, U=U), :U)
```
In my computer, the output of the above script is
```
false
flow for true.
```
Apparently, something is wrong. However, if the last line of the above script is replaced with the following
```julia
newexpand((t=t, U=U), :U)
```
The result is correct as expected:
```
false
flow for false.
```
In fact, even for
```julia
newexpand((t=t, μ=t, U=U), :U)
```
The code also works fine:
```
false
flow for false.
```
My Julia version is
```
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: AMD Ryzen 7 4800HS with Radeon Graphics
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, znver2)
```

### LINKED ISSUE #43082: Incorrect inference in MTK
Using Julia and MTK master:

```
using ModelingToolkit
using ModelingToolkit.BipartiteGraphs
graph = ModelingToolkit.BipartiteGraph(0, 0, Val(false))
f() = DiCMOBiGraph{true}(graph)
```

```
julia> @code_typed f()
CodeInfo(
1 ─ %1 = Main.graph::Any
│        (DiCMOBiGraph{true, I} where I)(%1)::Union{}
└──      unreachable
) => Union{}

julia> f()
Unreachable reached at 0x7f4edee39050
```

The inference of the constructor is incorrect. The function returns, causing the subsequent crash. I suspect an intersection issue of some sort as usual, but I will investigate further.

### PR DESCRIPTION
**Pull request by N5N3**
MWE
```julia
julia> A = Val{<:Union{Nothing, Val{true},Val{false}}}
Val{<:Union{Val{true}, Val{false}, Nothing}}

julia> B = Val{Val{B}} where {B}
Val{Val{B}} where B

julia> (Base.typeintersect(B, A))
Val{<:Union{Val{true}, Val{false}}}  # Val{Val{true}} before this PR
```

This PR makes sure `env` is restored between every 2 adjacent `intersect`s during `intersect_all`.
The valid env changes are tracked (and merged) with a new function `merge_env`.
Not sure this is the correct direction to fix the problem.

Close #41096.
Edit: looks like this PR also fixes #43082.

### CODE REVIEW

File: src/subtype.c [hunk 1]
```diff
@@ -3166,26 +3168,49 @@ static jl_value_t *intersect(jl_value_t *x, jl_value_t *y, jl_stenv_t *e, int pa
     return jl_bottom_type;
 }
 
+static int merge_env(jl_stenv_t *e, jl_value_t **root, jl_savedenv_t *se, int count) {
```
Comment (JeffBezanson) [inline src/subtype.c:3171]:
```suggestion
static int merge_env(jl_stenv_t *e, jl_value_t **root, jl_savedenv_t *se, int count)
{
```

**Comment by JeffBezanson** [comment]
Wow, this is great. Seems right to me, and other than that I go by the tests. I see one test was changed, but it's ok because the results are more conservative, so I'm happy with that. Thank you, this code is very tricky!

@nanosoldier `runtests(ALL, vs = ":master")`

**Review by JeffBezanson — APPROVED** [review]

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46350#issuecomment-1218542625) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/4c6f899_vs_0ada892/report.html).

**Comment by N5N3** [comment]
Results from PkgEval look like noise to me. Most of them are numeric/precompling errors.
Rerun the tests to confirm.

@nanosoldier `runtests(["AbstractAlgebra", "BokehServer", "CairoMakie", "ComplexPhasePortrait", "Contour", "CountdownNumbers", "DifferentiableTrajectoryOptimization", "FastMarching", "FilesystemDatastructures", "FlameGraphs", "Folds", "InformationInequalities", "IntervalLinearAlgebra", "LITS", "Leaflet", "OptimKit", "OteraEngine", "PCHIPInterpolation", "PProf", "Pfam", "Pidfile", "PointwiseKDEs", "PoreMatMod", "ProgressiveHedging", "ProxSDP", "ReinforcementLearningExperiments", "RetroCap", "RoundAndSwap", "StateSpaceModels", "StrBase", "SuiteSparseGraphBLAS", "SwagUI", "VisualRegressionTests", "Yunir"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46350#issuecomment-1219360922) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c951d94_vs_4af9674/report.html).

**Comment by N5N3** [comment]
@nanosoldier `runtests(["CairoMakie", "CountdownNumbers", "FlameGraphs", "Folds", "PProf", "Pidfile", "ReinforcementLearningExperiments", "RetroCap", "RoundAndSwap", "StrBase"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46350#issuecomment-1219523777) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c951d94_vs_1040707/report.html).

**Pull request merged**